File: src/loaders/cors.ts
import cors, { type CorsOptions } from 'cors';
import { appEnv } from '../lib/env.js';
import type { Application } from 'express';

const ALLOWED_ORIGINS = appEnv.VALID_ORIGINS.split(',');

export const corsOptions: CorsOptions = {
  origin: ALLOWED_ORIGINS,
  methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'],
  allowedHeaders: ['Authorization', 'Content-Type'],
  credentials: true
};

export default (app: Application): void => {
  app.use(cors(corsOptions));
};

File: internal/middleware/cors.go
package middleware

import (
	"net/http"
	"slices"

	"github.com/ccrsxx/api/internal/config"
)

func Cors(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestOrigin := r.Header.Get("Origin")

		allowedOrigins := config.Env().AllowedOrigins

		isOriginAllowed := slices.Contains(allowedOrigins, requestOrigin)

		if isOriginAllowed {
			w.Header().Set("Access-Control-Allow-Origin", requestOrigin)
			w.Header().Set("Access-Control-Allow-Credentials", "true")
			w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, PATCH, DELETE, OPTIONS")
			w.Header().Set("Access-Control-Allow-Headers", "Authorization, Content-Type")
		}

		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusNoContent)
			return
		}

		next.ServeHTTP(w, r)
	})
}
